src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -380,24 +623,71 @@ void CoreWorkerDirectTaskSubmitter::PushNormalTask(
       RAY_CHECK(scheduling_key_entry.active_workers.size() >= 1);
       RAY_CHECK(scheduling_key_entry.total_tasks_in_flight >= 1);
       scheduling_key_entry.total_tasks_in_flight--;
-    }
-    if (reply.worker_exiting()) {
-      // The worker is draining and will shutdown after it is done. Don't return
-      // it to the Raylet since that will kill it early.
-      absl::MutexLock lock(&mu_);
-      worker_to_lease_entry_.erase(addr);
-      auto &scheduling_key_entry = scheduling_key_entries_[scheduling_key];
-      scheduling_key_entry.active_workers.erase(addr);
-      if (scheduling_key_entry.CanDelete()) {
-        // We can safely remove the entry keyed by scheduling_key from the
-        // scheduling_key_entries_ hashmap.
-        scheduling_key_entries_.erase(scheduling_key);
+
+      if (reply.task_stolen()) {
+        RAY_CHECK(work_stealing_);
+        // add the task back to the queue
+        scheduling_key_entry.task_queue.push_back(task_spec);
+
+        // Obtain thief address
+        rpc::WorkerAddress thief_addr = rpc::WorkerAddress(reply.thief_addr());
+        RAY_LOG(DEBUG) << "Checking entry for thief " << thief_addr.worker_id
+                       << " still exists";
+        RAY_CHECK(worker_to_lease_entry_.find(thief_addr) !=
+                  worker_to_lease_entry_.end());
+        auto &thief_entry = worker_to_lease_entry_[thief_addr];
+
+        RAY_LOG(INFO) << "Record that thief " << thief_addr.worker_id
+                      << " has received one of the stolen tasks";
+        thief_entry.SetReceivedOneStolenTask();
+
+        if (!thief_entry.PipelineToWorkerFull(max_tasks_in_flight_per_worker_)) {
```

Comment:
I'm a bit worried about this new code to dispatch tasks, since it adds an additional branch that we need to think about.

We should try to keep the behavior similar to before. What if we tried something like this:
1. Add stolen tasks to the queue.
2. While the queue is non-empty, try to dispatch to the least full worker. This should be the thief.

It might sound similar, but this lets us always call one function to dispatch tasks to a worker (step 2), instead of having multiple ways of dispatching tasks.

Reply:
I actually copied this code from the `CoreWorkerDirectTaskSubmitter::SubmitTask` function so perhaps we should put the new code in a new method, and call that method in SubmitTask as well as in PushNormalTask? In that way, we don't duplicate code.
Also, to implement the functionality in point (2), should I just keep the current  `SchedulingKeyEntry::active_workers` flat_hash_set sorted by fullness of the workers' pipelines? In that way, the function above will automatically dispatch tasks to the least full worker